Source/DafnyCore/Backends/SinglePassCodeGenerator.cs

```diff
@@ -26,6 +26,13 @@ static class SinglePassCompilerExtensions {
   }
 
   public abstract class SinglePassCodeGenerator {
+
+    // Dafny names cannot start with "_". Hence, if an internal Dafny name is problematic in the target language,
+    // we can prefix it with "_".
+    // However, for backends such as Rust which need special internal fields, we want to clearly
+    // disambiguate between compiler-generated names and user-defined names, hence this prefix.
+    protected virtual string InternalFieldPrefix => "_i_";
```

Comment:
I don't fully understand. If Dafny user-defined names can not contain `_`, why do we need to do something special to distinguish them from compiler-generated names that contain `_` ?

Reply:
Dafny user-defined names can contain "_", it's just that they are escaped like this "__", so that "?" becomes "_q" and "'" becomes "_k". It makes the encoding bijective. Moreover, Dafny identifiers cannot start with "_", so anything starting with a "_" in generated code is never in collision with user-defined Dafny names, this is what the original InternalFieldPrefix is about (its value was "_"). So far so good, no issue.

However, if one wants to add variables in the code that are not defined in Dafny (e.g. for instrumentation), one incur the risk of colliding with user-defined Dafny variables. For example, if `the internal value of constant is a field prefixed with "_", a user writing this:
```dafny
class A {
  const s
}
```
will get a C# class like this:
```dfy
class A {
  private int _s;
  public int s => {
    // Compute _s if not computed.
    return _s;
  }
}
```

Now let's say I want to write a field, like "time" for instrumentation. If I prefix it with "_", like "_time", it might collide with a user-defined constant name "time".
Moreover, if one wants to keep an encoding where "a_b" is translated to "a_b" (e.g. if an identifier is Rust-friendly in my case, we don't change it), then there is no possibility to avoid collisions with an internal prefix like "_" for any field and variable name. For example, let's say I add a private field "_a_b", then a user defining a constant "a_b" will collide with my private field.

To avoid that issue, it suffices to increase the length of the internal prefix, which I did here; Now I can add any private field that starts with `_set_` for example (which I do for constructor variable initialization later) and I know it won't collide with any user-defined private field. In some sense, it serves as a namespace.
I could as well have put it with two underscores, "__", I'm pretty sure it would work too, but I haven't tried; At least this prefix is easily recognizable and discoverable in case of bugs.

By the way, In this PR I'm putting the internal prefix for all compilers to "_i_" but I could put it back to "_" for all of them except the Rust compiler if we wanted to. The fact however that it captures all the places of where this prefix is used makes it more maintainable in the future, e.g. in case we want to generate more idiomatic output in the future for other compilers.